server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -617,6 +629,18 @@ static ClusterState innerRemoveIndexTemplateV2(ClusterState currentState, String
         return ClusterState.builder(currentState).metadata(metadata).build();
     }
 
+    static Set<String> dataStreamsUsingTemplate(final ClusterState state, final String templateName) {
+        final Set<String> dataStreams = state.metadata().dataStreams().keySet();
```

Comment:
pico-optimization: `dataStreams` not needed until after `templateName` is confirmed to exist